STORY-6683: support UTF-16LE and UTF-16BE in guaranteed_utf8_string

ColinDKelley opened the pull request:
https://ringrevenue.atlassian.net/browse/STORY-6683

@primiti  Can you please review?

ColinDKelley committed:
STORY-6683: rubocop -a

ColinDKelley committed:
STORY-6683: expose normalize_string class method with separate keyword options

ColinDKelley committed:
STORY-6683: refactor tests to use normalize_string primarily and to explicitly test each optional conversion with true/false

ColinDKelley committed:
STORY-6683: add support for UTF-16LE and UTF-16BE, including CP1252 incorrectly encoded inside those

ColinDKelley committed:
STORY-6683: rubocop -a

ColinDKelley commented on lib/invoca/utils/guaranteed_utf8_string.rb:
```diff
@@ -1,68 +1,103 @@
 
       # chosen because this is a 1-byte ASCII character that is not used in any of the popular escaping systems: XML, HTML, HTTP URIs, HTTP Form Post, JSON
-      REPLACE_CHARACTER = '~' unless defined?(REPLACE_CHARACTER)
+      REPLACE_CHARACTER = '~'
```
Now that I know the cause of those warnings, I'm getting rid of this lame work-around!

ColinDKelley committed:
STORY-6683: add .rubocop.yml and add .rubocop-http* to .gitignore

ColinDKelley committed:
STORY-6683: make normalize_utf16: optional

primiti commented on lib/invoca/utils/guaranteed_utf8_string.rb:
```diff
@@ -1,68 +1,107 @@
-        if !str.valid_encoding?
-          cp1252_to_utf_8(str)
+      class << self
+        def normalize_string(orig_string, normalize_utf16: true,
```
Very minor, but do you want to put the second arg on its on line as well?

```suggestion
        def normalize_string(orig_string, 
                                           normalize_utf16: true,
```

ColinDKelley committed:
STORY-6683: allow normalize_cp1252: false to be used with normalize_utf16: true (making them orthogonal)

ColinDKelley commented:
@primiti  Thanks for the comments. They are addressed here: https://github.com/Invoca/invoca-utils/pull/3/commits/9d9b249015c2c1af1a6e47e188c4c5daf815aadf

(Also I addressed a Rubocop comment that the complexity of that method was getting too high. They were right. I left the part that normalized the argument and moved all the rest into a private method.)

ColinDKelley committed:
STORY-6683: extract constant and use 4-digit Unicode code point format
